open-feature/open-feature-operator
# feat: Custom labels and annotations for namespace.

## This PR
Allows custom labels and annotations to be added to a namespace. It also makes it possible to simply disable namespace creation for clusters where ownership of the namespace needs to lie elsewhere for policy or structural reasons.

### Related Issues
Re-implements https://github.com/open-feature/open-feature-operator/pull/678

### Notes
#### Note 1
Fully rendered `namespace.yaml` after running `make helm-package` (Extracted from `/charts/open-feature-operator-v0.7.1.tgz` using `tar -xvf open-feature-operator-v0.7.1.tgz`)

```
# Only deploy the namespace if the default is being used (helm install should fail if the namespace isnt present)
# when one is defined with -n
{{- if and (eq (include "chart.namespace" .) .Values.defaultNamespace) .Values.namespace.create }}
apiVersion: v1
kind: Namespace
metadata:
  name: '{{ include "chart.namespace" . }}'
  labels:
    control-plane: controller-manager
    {{- range $key, $value := $.Values.namespace.labels }}
    {{ $key }}: {{ $value | quote }}
    {{- end }}
  annotations:
    {{- range $key, $value := $.Values.namespace.annotations }}
    {{ $key }}: {{ $value | quote }}
    {{- end }}
{{ end }}
```
#### Note 2
This does not need to utilize any of the special directives of [strip-kustomize-helm.sh](https://github.com/open-feature/open-feature-operator/blob/main/.github/scripts/strip-kustomize-helm.sh) because it is not referenced in the [helm overlay kustomization](https://github.com/open-feature/open-feature-operator/blob/main/config/overlays/helm/kustomization.yaml)

#### Note 3
I side-stepped the issue that https://github.com/open-feature/open-feature-operator/pull/678 has with the confusion between `namespace.name` and `defaultNamespace` by simply leaving `defaultNamespace` in place and not adding another name field. It is slightly awkward because now we have a `namespace` section and a standalone `defaultNamespace` variable. However, I don't believe it's worth breaking backwards compatibility over this.

### Follow-up Tasks
- Need to validate https://github.com/open-feature/open-feature-operator/issues/683. I believe this PR does fix it.

### How to test
1. Add labels and annotations to the values.
2. Deploy the helm chart with `namespace.create: true` set
3. Validate that the labels and annotations appear on the namespace.
4. Set defaultNamespace to `null` and test with create to be true/false (if false, create namespace externally from the helm chart)
